PaletteEdit: temporary custom gradient not saving

ramonjd opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/56865

## What?
Refactors `isTemporaryElement` test in the `PaletteEdit` component.

## Why?
Gradient elements contain both color and gradient properties, therefore they'll always return true for the `isTemporaryElement` test if the color property is default (`#000`).

## How?
Ensuring we test for `gradient` values first.

## Testing Instructions

1. Kick on over to the site editor and open up the global styles panel.
2. Create a new custom gradient in Colors > Palette > Gradient (ensuring to modify the color at least)
3. Click "Done"
4. The new custom gradient should persist

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

### Before
https://github.com/WordPress/gutenberg/assets/6458278/ecd3e33d-4270-4b8e-82fa-1a45507990f6


### After
https://github.com/WordPress/gutenberg/assets/6458278/c1c2e4c5-a0f0-4407-93df-7a09ce2d39f0

ramonjd committed:
Gradient elements contain both color and gradient properties, therefore they'll always return true for this test if the color property is default (#000)

ramonjd committed:
CHANGELOG.md

tellthemachines reviewed (commented):
Thanks for the PR! 

It does fix the inability to save a custom gradient with the default colour, but now it's not possible to save the default gradient with a custom name. I wonder if we shouldn't be checking that _both_ name and gradient are set to default, and if either one has changed assume the change is deliberate and save?

One thing I noticed while trying to reproduce the issue on trunk is that it's only impossible to save a gradient with the default name if there are no other gradients saved already. After successfully saving one, I'm able to save however many I want with the default name. Not sure why this happens though.

ramonjd committed:
isTemporaryElement should return false by default so that it will catch changes to the slug
Added tests

ramonjd commented:
> now it's not possible to save the default gradient with a custom name.

I've fixed that I think, and added some tests. I was returning the wrong default value from `isTemporaryElement`.

Working on this PR highlighted another logical error in trunk: that the regex `^${ slugPrefix }color-([\\d]+)$` will catch all `color-${ anyNumber }`, so if a user edits the number on the default name and doesn't update the color value, it'll be treated as a "temporary" color, and wiped.

https://github.com/WordPress/gutenberg/assets/6458278/bfc31a97-a4e4-44d3-a29e-c8bad2e74867

Might require a bit of a refactor so I'd say unrelated to this bug, which fixes the bug in the color/gradient conditional test order.

ramonjd commented:
> Might require a bit of a refactor so I'd say unrelated to this bug, which fixes the bug in the color/gradient conditional test order.

Have WIP PR here as a follow up to this one:

- https://github.com/WordPress/gutenberg/pull/56932

andrewserong reviewed (approved):
Works nicely for me, and is a definite improvement over what we have on trunk. Thanks for adding the tests, too!

Saving custom gradients is working now, and saving single colors is still working:

https://github.com/WordPress/gutenberg/assets/14988353/a38ecfec-6d52-4112-a0d5-1f367a8068c3

LGTM! ✨

ramonjd commented:
Thanks @andrewserong!

ciampo commented on packages/components/src/palette-edit/index.tsx:
```diff
@@ -261,27 +261,30 @@ function Option< T extends Color | Gradient >( {
 	slugPrefix: string,
 	{ slug, color, gradient }: Color | Gradient
-) {
+): Boolean {
```
Minor comment, but [we should use `boolean` instead of `Boolean`](https://stackoverflow.com/questions/64443288/what-is-the-difference-between-boolean-and-boolean-in-typescript)

ramonjd replied:
Oh! Thanks. I'll fix in an upcoming, related PR 👍🏻

andrewserong replied:
Great catch, thanks Marco!

ciampo reviewed (commented):
Thank you for working on this (and for adding tests!!) I've left a small comment, but otherwise this looks great!
